org/postgresql/jdbc2/StatementFinalizer.java

```diff
@@ -0,0 +1,70 @@
+package org.postgresql.jdbc2;
+
+import java.sql.Statement;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+/**
+ * This class closes the statements that were not closed by the application.
+ * The idea of using a separate class for handling finalization is to avoid usage
+ * of finalizer in the base connection class since finalizers retain garbage longer
+ * than regular objects and finalization is not scalable yet in all the JMVs.
+ */
+/* package */ class StatementFinalizer
+{
+    /**
+     * The updater is used to implement compare and swap on the {@link #statement} field.
+     * We could use {@link java.util.concurrent.atomic.AtomicReference}, however it would
+     * increase heap size retained by the finalizer.
+     */
+    private AtomicReferenceFieldUpdater<StatementFinalizer, Statement> STATEMENT_UPDATER =
```

Comment:
Not sure if that's a yes or a no. Let me rephrase ...

I meant that the `AtomicReferenceFieldUpdater` should be static. Otherwise you're creating a new one for every Statement object. The whole point of using it with the volatile field, instead of using an `AtomicReference` , is to avoid the per-object additional reference.